File: src/config/loader.ts
/**
 * Configuration file loader
 * Handles reading and writing the sage configuration file
 */

import { readFile, writeFile, mkdir, access } from 'node:fs/promises';
import { homedir } from 'node:os';
import { join } from 'node:path';
import type { UserConfig } from '../types/index.js';
import { DEFAULT_CONFIG } from '../types/config.js';
import { validateCalendarSources } from './validation.js';
import { MeetingIntelligenceConfigSchema, SlackIntegrationConfigSchema } from '../types/pipeline-config.js';

const SAGE_DIR = '.sage';
const CONFIG_FILE = 'config.json';

export class ConfigLoader {
  /**
   * Get the path to the sage configuration directory
   */
  static getConfigDir(): string {
    return join(homedir(), SAGE_DIR);
  }

  /**
   * Get the path to the configuration file
   */
  static getConfigPath(): string {
    return join(this.getConfigDir(), CONFIG_FILE);
  }

  /**
   * Check if the configuration file exists
   */
  static async exists(): Promise<boolean> {
    try {
      await access(this.getConfigPath());
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Load the configuration from disk
   * Throws if the file doesn't exist or is invalid
   */
  static async load(): Promise<UserConfig> {
    const configPath = this.getConfigPath();

    try {
      const content = await readFile(configPath, 'utf-8');
      const parsed = JSON.parse(content) as UserConfig;

      // Validate basic structure
      if (!parsed.version || !parsed.user) {
        throw new Error('Invalid configuration file structure');
      }

      // Migrate config if calendar.sources is missing
      let migrated = false;
      if (!parsed.calendar.sources) {
        // Deep copy to avoid reference sharing
        parsed.calendar.sources = JSON.parse(
          JSON.stringify(DEFAULT_CONFIG.calendar.sources)
        );
        migrated = true;
      }

      // Migrate config if meetingIntelligence is missing
      if (!parsed.meetingIntelligence) {
        parsed.meetingIntelligence = MeetingIntelligenceConfigSchema.parse({});
        migrated = true;
      }

      // Migrate integrations: backfill any missing sub-fields from DEFAULT_CONFIG.
      // Older configs (and test fixtures) may omit googleCalendar/notion/appleReminders/slack;
      // downstream adapters access these unconditionally, so we ensure shape parity here.
      if (!parsed.integrations) {
        parsed.integrations = JSON.parse(JSON.stringify(DEFAULT_CONFIG.integrations));
        migrated = true;
      } else {
        if (!parsed.integrations.appleReminders) {
          parsed.integrations.appleReminders = JSON.parse(
            JSON.stringify(DEFAULT_CONFIG.integrations.appleReminders)
          );
          migrated = true;
        }
        if (!parsed.integrations.notion) {
          parsed.integrations.notion = JSON.parse(
            JSON.stringify(DEFAULT_CONFIG.integrations.notion)
          );
          migrated = true;
        }
        if (!parsed.integrations.googleCalendar) {
          parsed.integrations.googleCalendar = JSON.parse(
            JSON.stringify(DEFAULT_CONFIG.integrations.googleCalendar)
          );
          migrated = true;
        }
        if (!parsed.integrations.slack) {
          parsed.integrations.slack = SlackIntegrationConfigSchema.parse({});
          migrated = true;
        }
      }

      // Backfill autonomy block when missing. Default to Tier 1 across all
      // known write tools so existing configs become safe-by-default after
      // upgrade rather than silently auto-executing every mutation.
      if (!parsed.autonomy) {
        parsed.autonomy = JSON.parse(JSON.stringify(DEFAULT_CONFIG.autonomy));
        migrated = true;
      }

      // Validate calendar.sources if present
      if (parsed.calendar.sources) {
        const validation = validateCalendarSources(parsed.calendar.sources);
        if (!validation.success) {
          throw new Error(
            `Invalid calendar sources configuration: ${validation.error?.message}`
          );
        }
      }

      // Save migrated config
      if (migrated) {
        await this.save(parsed);
      }

      return parsed;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new Error('Configuration file not found');
      }
      throw error;
    }
  }

  /**
   * Save the configuration to disk
   */
  static async save(config: UserConfig): Promise<void> {
    const configDir = this.getConfigDir();
    const configPath = this.getConfigPath();

    // Validate calendar.sources (required)
    if (!config.calendar.sources) {
      throw new Error('Missing required field: calendar.sources');
    }

    const validation = validateCalendarSources(config.calendar.sources);
    if (!validation.success) {
      throw new Error(
        `Invalid calendar sources configuration: ${validation.error?.message}`
      );
    }

    // Ensure the directory exists
    await mkdir(configDir, { recursive: true });

    // Update timestamp
    const updatedConfig: UserConfig = {
      ...config,
      lastUpdated: new Date().toISOString(),
    };

    // Write the file with pretty formatting
    await writeFile(configPath, JSON.stringify(updatedConfig, null, 2), 'utf-8');
  }

  /**
   * Load the configuration, or create a default one if it doesn't exist
   */
  static async loadOrCreate(): Promise<UserConfig> {
    try {
      return await this.load();
    } catch {
      const config = this.getDefaultConfig();
      await this.save(config);
      return config;
    }
  }

  /**
   * Get the default configuration
   */
  static getDefaultConfig(): UserConfig {
    // Deep copy DEFAULT_CONFIG to avoid reference sharing
    const config = JSON.parse(JSON.stringify(DEFAULT_CONFIG)) as UserConfig;
    return {
      ...config,
      createdAt: new Date().toISOString(),
      lastUpdated: new Date().toISOString(),
    };
  }

  /**
   * Merge partial updates into an existing configuration
   */
  static mergeConfig(base: UserConfig, updates: Partial<UserConfig>): UserConfig {
    return {
      ...base,
      ...updates,
      lastUpdated: new Date().toISOString(),
    };
  }
}

File: rust/crates/sage-config/src/loader.rs
//! Config loading/saving with lazy migration. Mirrors `src/config/loader.ts`.
//!
//! Migration operates on the parsed JSON `Value` (like the TS code mutating the
//! parsed object) before deserialization, so old/partial configs that omit
//! `calendar.sources`, `meetingIntelligence`, `integrations.*`, or `autonomy`
//! load successfully and are re-saved with the backfilled defaults.

use crate::default::default_config;
use crate::types::UserConfig;
use crate::validation::{validate_calendar_sources, ValidationError};
use chrono::{SecondsFormat, Utc};
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, thiserror::Error)]
pub enum ConfigError {
    #[error("home directory not found")]
    NoHomeDir,
    #[error("configuration file not found")]
    NotFound,
    #[error("invalid configuration file structure: {0}")]
    InvalidStructure(String),
    #[error(transparent)]
    InvalidCalendarSources(#[from] ValidationError),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("json error: {0}")]
    Json(#[from] serde_json::Error),
}

/// `~/.sage`
pub fn config_dir() -> Result<PathBuf, ConfigError> {
    Ok(dirs::home_dir()
        .ok_or(ConfigError::NoHomeDir)?
        .join(".sage"))
}

/// `~/.sage/config.json`
pub fn config_path() -> Result<PathBuf, ConfigError> {
    Ok(config_dir()?.join("config.json"))
}

/// Whether the config file exists.
pub fn exists() -> bool {
    config_path().map(|p| p.exists()).unwrap_or(false)
}

/// Backfill missing optional sections from the defaults, mutating `value` in
/// place. Returns `true` if anything was added. Pure (no I/O) for testability.
pub(crate) fn migrate(value: &mut Value) -> bool {
    let defaults = serde_json::to_value(default_config()).expect("default_config serializes");
    let mut migrated = false;

    let Some(obj) = value.as_object_mut() else {
        return false;
    };

    // calendar.sources
    if let Some(cal) = obj.get_mut("calendar").and_then(Value::as_object_mut) {
        if !cal.contains_key("sources") {
            cal.insert("sources".into(), defaults["calendar"]["sources"].clone());
            migrated = true;
        }
    }

    // meetingIntelligence
    if !obj.contains_key("meetingIntelligence") {
        obj.insert(
            "meetingIntelligence".into(),
            defaults["meetingIntelligence"].clone(),
        );
        migrated = true;
    }

    // integrations (whole block, then per-sub-field backfill)
    if !obj.contains_key("integrations") {
        obj.insert("integrations".into(), defaults["integrations"].clone());
        migrated = true;
    } else if let Some(integ) = obj.get_mut("integrations").and_then(Value::as_object_mut) {
        for key in ["appleReminders", "notion", "googleCalendar", "slack"] {
            if !integ.contains_key(key) {
                integ.insert(key.into(), defaults["integrations"][key].clone());
                migrated = true;
            }
        }
    }

    // autonomy (safe-by-default: backfills the Tier-1 matrix)
    if !obj.contains_key("autonomy") {
        obj.insert("autonomy".into(), defaults["autonomy"].clone());
        migrated = true;
    }

    migrated
}

/// Load and (lazily migrate) the config from `~/.sage/config.json`.
pub fn load() -> Result<UserConfig, ConfigError> {
    let path = config_path()?;
    let content = match fs::read_to_string(&path) {
        Ok(c) => c,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Err(ConfigError::NotFound),
        Err(e) => return Err(e.into()),
    };

    let mut value: Value = serde_json::from_str(&content)?;

    // Basic structure check (TS: `parsed.version && parsed.user`).
    let has_version = value
        .get("version")
        .and_then(Value::as_str)
        .is_some_and(|s| !s.is_empty());
    if !has_version || value.get("user").is_none() {
        return Err(ConfigError::InvalidStructure(
            "missing version or user".into(),
        ));
    }

    let migrated = migrate(&mut value);
    let config: UserConfig = serde_json::from_value(value)?;

    if let Some(sources) = &config.calendar.sources {
        validate_calendar_sources(sources)?;
    }

    if migrated {
        save(&config)?;
    }
    Ok(config)
}

/// Persist the config, validating `calendar.sources` and stamping `lastUpdated`.
pub fn save(config: &UserConfig) -> Result<(), ConfigError> {
    let sources = config.calendar.sources.as_ref().ok_or_else(|| {
        ConfigError::InvalidStructure("missing required field: calendar.sources".into())
    })?;
    validate_calendar_sources(sources)?;

    fs::create_dir_all(config_dir()?)?;

    let mut to_write = config.clone();
    to_write.last_updated = Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true);

    fs::write(config_path()?, serde_json::to_string_pretty(&to_write)?)?;
    Ok(())
}

/// A fresh default config (timestamps set to now). Mirrors `getDefaultConfig`.
pub fn get_default_config() -> UserConfig {
    default_config()
}

/// Load the config, creating a default one if it doesn't exist or is invalid.
pub fn load_or_create() -> Result<UserConfig, ConfigError> {
    match load() {
        Ok(config) => Ok(config),
        Err(_) => {
            let config = get_default_config();
            save(&config)?;
            Ok(config)
        }
    }
}
